### PR TITLE
Add timezones support to promql

### PR DESCRIPTION
**Pull request by sylr**
This function aims to solve problems for those who need to account for DST in their queries.

### CODE REVIEW

**Comment by juliusv** [comment]
Thanks! Others always have more to say about timezones than me, but I do remember a previous conversation about just this:

* https://github.com/prometheus/prometheus/issues/8872
* https://github.com/prometheus/prometheus/pull/8871#issuecomment-850584432

That second comment from Björn about the edge cases that a simple offset function wouldn't cover is especially relevant to think about.

**Comment by beorn7** [comment]
Also a lot of discussion can be found on this design doc: https://docs.google.com/document/d/1xfw1Lb1GIRZB_-4iFVGkgwnpwuBemWfxYqFdBm7APsE/edit#heading=h.bupciudrwmna

tl;dr: This is not as easy as it looks.

That's just very generally about the state of this problem. I haven't looked into this PR in particular, but I'll put this on my review queue for a proper review. (Sadly, my backlog is quite significant here.)

**Comment by sylr** [comment]
Pasted_Image_13_01_2022_19_28

**Comment by sylr** [comment]
Quite frankly I fail to see what is "not that easy".

I don't claim to see all the possible timezone problems one can have but as a +5 years old daily prometheus user, give me this or an equivalent (`hour(vector(time()), "Europe/Paris")`, `hour(vector(time("Europe/Paris")))` ... etc) and https://github.com/prometheus/alertmanager/pull/2782 and you would have solved 100% of my prometheus timezone related problems.

**Comment by juliusv** [comment]
@sylr Yeah, if I understand correctly, because you provide a full `time_tz()` function and not a pure offset function, your approach doesn't actually have the DST switch problem as described in https://github.com/prometheus/prometheus/pull/8871#issuecomment-850584432. As far as I can see, @beorn7's design doc about time handling doesn't actually say much about time zones, but there's a bit more general stuff in that PR discussion.

While I'm not the timezone guy, there are still at least some things I'd like a reviewer to take into account:

* Do we want to introduce a PromQL function that does not work on all platforms (or alternatively, can it be made to work on all)?
* When something is officially part of PromQL, all other projects and vendors who want to use the "Prometheus Compatible" mark will need to implement this. Currently all compatible systems are written in Go (or have a Go layer that creates comaptibility), but does timezone handling introduce any unforeseenly large burden on other implementors?
* Time zone definitions around the world change regularly (laws around it are being passed all the time). Most people will not be fully up to date with the very latest time zone database changes. Especially given the point above, what does this mean for compatibility testing? Or do we just exempt that area from compatibility for the most part?

That said, if those concerns are at least acknowledged and deemed fine, I would be happy if timezone functionality (such as this or another viable alternative) would make it into Prometheus, so that users can stop using weird workarounds such as the one mentioned in https://github.com/prometheus/prometheus/pull/8871#issuecomment-850652786.

**Comment by sylr** [comment]
> Do we want to introduce a PromQL function that does not work on all platforms (or alternatively, can it be made to work on all)?

Here are the possible choices I see:
- Disallow timezone related functions on Windows
- Allow them knowing that they are currently broken
- Build a Windows version with the go build tag `timetzdata` so that windows build get working timezones but embedded in the binary so not coming from the OS.

The status quo of not having timezone support in PromQL is not an option (according to me).

> When something is officially part of PromQL, all other projects and vendors who want to use the "Prometheus Compatible" mark will need to implement this. Currently all compatible systems are written in Go (or have a Go layer that creates comaptibility), but does timezone handling introduce any unforeseenly large burden on other implementors?

Honestly I don't know. However, I would be surprised to find an all-purpose language without the basic functions to do timezone diffs in its standard library.

> Time zone definitions around the world change regularly (laws around it are being passed all the time). Most people will not be fully up to date with the very latest time zone database changes. Especially given the point above, what does this mean for compatibility testing? Or do we just exempt that area from compatibility for the most part?

Timezone definitions shouldn't be anyone else's responsibility but the OS.

**Comment by juliusv** [comment]
Yeah, fair views. What do people thinking about adding `time_tz()` behind an experimental feature flag (`--enable-feature=time-tz`) maybe for the time being? That way we have some time to decide on the Windows and compatibility questions before making it an official part of PromQL, and at the same time if we decide on a different way of approaching this functionality, the door would still be open to that.

**Comment by sylr** [comment]
> What do people thinking about adding time_tz() behind an experimental feature flag (--enable-feature=time-tz) maybe for the time being?

I'm all for it if it can get things moving on the matter.

**Comment by beorn7** [comment]
The discussion about timezone support has happened at many places already:
- #8871
- #8872
- [Design doc draft](https://docs.google.com/document/d/1xfw1Lb1GIRZB_-4iFVGkgwnpwuBemWfxYqFdBm7APsE/edit#heading=h.bupciudrwmna)
- [At the dev summit](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit?pli=1#heading=h.eohahn77vze0) (search for "Support timezones in Prometheus").
- Some parts of this [brainstorming doc about timestamps and durations].
- (And probably a few more I forgot.)

All of these demonstrate the problems, but also provide a lot of ideas and insights how to approach them.

I don't think it's helpful to start another lengthy discussion in this PR and disregard everything that has been discussed before. (Stating that "quite frankly I fail to see what is 'not that easy'" won't convince the participants of above discussions that they were all wrong all the time.)

I think the next step should be to finalize a design doc (by amending the existing one or rewriting it – the existing one was around the basic idea of providing a timezone on the command line, which hit a lot of resistance, so that approach would just be one of the possible alternatives to be presented in the design doc).

Thanks to feature flags, we can certainly experiment with features where we are unsure if they will work or if they will be useful. However, in this case, we already know about concrete problems, and various ideas have been expressed how to address them. Those ideas need to be discussed rather than disregarded. Simply implementing a feature while "failing to see the problem" isn't helping.

**Comment by beorn7** [comment]
Naming nit: `time_tz` is a bad name as the argument is a location, not a timezone.

`time.LoadLocation` returns the (current) TZ for the provided location. The TZ for "Europe/Paris" is either "CEST" or "CET", depending on the time of the year. Or in other words: If you know the timezone, you already know if you are affected by so-called "DST" or not.

The name should be something like `time_for_location`.

**Comment by roidelapluie** [comment]
We would need to have the embedded TZ for windows (via build flags) + have a dedicated flag (for all os'es) to pass a recent tzdata file. Somehow it would be great to expose the tz version as metric.

**Comment by beorn7** [comment]
About how to provide the TZ data: The [design doc]() already contains an interesting discussion with a conclusion in the comments. Search for "Should we provide the embedded tzdata for windows users?" and also read the comments there.

**Comment by sylr** [comment]
How would you propagate the feature flag value to `github.com/prometheus/prometheus/promql/parser` ?

**Comment by roidelapluie** [comment]
> Naming nit: `time_tz` is a bad name as the argument is a location, not a timezone.
> 
> `time.LoadLocation` returns the (current) TZ for the provided location. The TZ for "Europe/Paris" is either "CEST" or "CET", depending on the time of the year. Or in other words: If you know the timezone, you already know if you are affected by so-called "DST" or not.
> 
> The name should be something like `time_for_location`.

I disagree with this. As per wikipedia, within the tz database, a time zone is any national region where local clocks have all agreed since 1970. In general, Timezone is way more clear than location. https://en.wikipedia.org/wiki/Tz_database#File_formats

File: cmd/prometheus/main.go [hunk 1]
```diff
@@ -185,6 +186,18 @@ func (c *flagConfig) setFeatureListOptions(logger log.Logger) error {
 			case "agent":
 				agentMode = true
 				level.Info(logger).Log("msg", "Experimental agent mode enabled.")
+			case "timezones":
+				if runtime.GOOS == "windows" {
+					zoneinfo := os.Getenv("ZONEINFO")
+					if len(zoneinfo) == 0 {
+						return errors.New("timezones support on Windows requires env var ZONEINFO to be supplied")
```
Comment (roidelapluie) [inline cmd/prometheus/main.go:193]: We should not take configuration from environment variable, and custom timezone support could be offered for other os'es as well.

**Comment by roidelapluie** [comment]
I think it would be cleaner to have the exception handled by the engine rather than by the parser.

File: cmd/prometheus/main.go [hunk 2]
```diff
@@ -185,6 +186,18 @@ func (c *flagConfig) setFeatureListOptions(logger log.Logger) error {
 			case "agent":
 				agentMode = true
 				level.Info(logger).Log("msg", "Experimental agent mode enabled.")
+			case "timezones":
+				if runtime.GOOS == "windows" {
```
Comment (roidelapluie) [inline cmd/prometheus/main.go:190]: Parsing feature flags should never fail, and I don't think we should have this check. I think we should rely on Go's default behaviour and err in LoadLocation (I guess that's what it does by default).

**Comment by beorn7** [comment]
WRT https://en.wikipedia.org/wiki/Tz_database#File_formats : Indeed, the TZ database uses locations to name timezones. However, https://pkg.go.dev/time has a usage that corresponds to what I had said. Therefore, "tz" or "timezone" is ambiguous outside of the immediate context of the TZ database. "location" is not ambiguous and I would prefer if we used that.

**Comment by roidelapluie** [comment]
It seems that this is still using parser, while other experimental features like @ fail at the engine side. Additionally, engine already has a mechanism for feature enabling (see EngineOpts).

**Comment by sylr** [comment]
You want to overwrite the Parser's Functions from the engine ?

**Comment by roidelapluie** [comment]
I think we should maybe have a second FunctionCalls in the engine that can be used to process this. The initial Functioncalls would fail if there is at tz passed.

**Comment by beorn7** [comment]
Thanks for all the work you are investing here. However, I'm concerned that we go deep into details here without having an agreed-upon design doc. We are discussing general aspects mixed up with implementation details here, which makes it hard to follow for a wider audience. Plus, a PR is just not the right form to make important design decisions.

Ideas seem to take shape here, so I guess it won't be that hard to write them down in a design doc, with their reasoning, so that the can be presented to the developer community, discussed in a more visible and more focused way, get buy-in from the community as a whole, and ideally get final agreement at a dev summit.

All of this reminds me that we should really better document our proposal process, cf. https://github.com/prometheus/prometheus/pull/8986

**Comment by valyala** [comment]
How about much simpler solution - to add [timezone_offset](https://docs.victoriametrics.com/MetricsQL.html#timezone_offset) function, which would return the offset in seconds for the specified timezone. For example, `timezone_offset("America/Los_Angeles")` would return `-28800`, e.g. `-8h`. Then this function can be freely combined with other time-related functions. For example, `floor((time() + timezone_offset("America/Los_Angeles")) / 3600) % 24` would return the hour at `America/Los_Angeles` timezone.

An additional benefit for having a separate `timezone_offset()` function is that it can be combined with time series values, which store timestamps. For example, `prometheus_config_last_reload_success_timestamp_seconds + timezone_offset("America/los_Angeles")` would return the last successful config reload time in seconds in `America/Los_Angeles` time zone.

This feature is already implemented in VictoriaMetrics, so you can [play with it there](https://play.victoriametrics.com/select/accounting/1/6a716b0f-38bc-4856-90ce-448fd713e3fe/prometheus/graph/?g0.expr=floor((time()%20%2B%20timezone_offset(%22America%2FLos_Angeles%22))%20%2F%203600)%20%25%2024&g0.range_input=2d&g0.end_input=2023-02-03T23%3A55%3A44&g0.relative_time=last_2_days).

Additionally, it would be great if Prometheus would support time-related constants anywhere in the query. For example, `8h` would be automatically converted to `8*3600`. This will simplify time-related calculations. For example, the following query would return the average rate for samples' scraping over the last hour:

```metricsql
sum(sum_over_time(scrape_samples_scraped[1h])) / 1h
```

Or, if you know needed timezone offset, you could write something like the following `time() + 5h30m` for [Indian standard time](https://en.wikipedia.org/wiki/Indian_Standard_Time). This feature is also supported by VictoriaMetrics, so you can [play with it there](https://play.victoriametrics.com/select/accounting/1/6a716b0f-38bc-4856-90ce-448fd713e3fe/prometheus/graph/?g0.expr=time()%20%2B%205h30m&g0.range_input=2d&g0.end_input=2023-02-04T00%3A02%3A50&g0.relative_time=last_2_days).

**Comment by roidelapluie** [comment]
The solution of using the timezone_offset function to calculate time in a specific time zone may not be suitable for all use cases, as it may not be compatible with metrics collected prior to Daylight Saving Time (DST) changes. This is because the timezone_offset function provides the offset for the current query time and not the metric time. As a result, if a metric was collected prior to a DST change, the calculation of time in a specific time zone using timezone_offset would be incorrect.

**Comment by juliusv** [comment]
Aside: Issues with `timezone_offset()` have already been discussed above and in other places by @beorn7, such as https://github.com/prometheus/prometheus/pull/10163#issuecomment-1013148678 and https://github.com/prometheus/prometheus/pull/8871#issuecomment-850584432.

**Comment by beorn7** [comment]
Thanks, @juliusv, for linking my relevant ramblings. :)

WRT making all durations numbers and vice versa, there is already https://github.com/prometheus/prometheus/pull/9138 . It's almost done, IMHO, just somebody has to shepherd it towards conclusion.

**Comment by bboreham** [comment]
As said at https://github.com/prometheus/prometheus/pull/10163#issuecomment-1013125293
> I think the next step should be to finalize a design doc (by amending the existing one or rewriting it 

Given there has been a lot of debate and no progress, we decided at the bug-scrub meeting to close this PR until a design is agreed.
